sockeye/optimizers.py

```diff
@@ -0,0 +1,244 @@
+
+    This version allows:
+        * Using validation checkpoint loss in addition to training batch loss.
+        * Using Adam or Nesterov Adam (Nadam) as the base algorithm
+
+    Eve does not currently support rescaling gradients, clipping gradients, or weight decay.
+
+    :param learning_rate: The initial learning rate.
+    :param beta1: Exponential decay rate for the first moment estimates.
+    :param beta2: Exponential decay rate for the second moment estimates.
+    :param beta3_batch: Exponential decay rate for batch objective relative change.
+    :param beta3_checkpoint: Exponential decay rate for checkpoint objective relative change.
+    :param epsilon: Small value to avoid division by 0.
+    :param k_lo: Lower threshold for relative change.
+    :param k_hi: Upper threshold for relative change.
+    :param use_batch_objective: Incorporate batch objective (can use both).
+    :param use_checkpoint_objective: Incorporate checkpoint objective (can use both).
+    :param use_nesterov_momentum: Use Nesterov-accelerated adaptive moment estimation (update rules
+                                  used by "Nadam" optimizer).
+    """
+    def __init__(self,
+                 learning_rate: float = 0.001,
+                 beta1: float = 0.99,
+                 beta2: float = 0.999,
+                 beta3_batch: float = 0.999,
+                 beta3_checkpoint: float = 0.,
+                 epsilon: float = 1e-8,
+                 k_lo: float = 0.1,
+                 k_hi: float = 10,
+                 schedule_decay: float = 0.004,
+                 use_batch_objective: bool = True,
+                 use_checkpoint_objective: bool = False,
+                 use_nesterov_momentum: bool = True,
+                 **kwargs) -> None:
+        check_condition(any((use_batch_objective, use_checkpoint_objective)),
+                        "Must use at least one of: batch objective, checkpoint objective")
+        super().__init__(learning_rate=learning_rate, **kwargs)
+        self.beta1 = beta1
+        self.beta2 = beta2
+        self.beta3_batch = beta3_batch
+        self.beta3_checkpoint = beta3_checkpoint
+        self.epsilon = epsilon
+        self.k_lo = k_lo
+        self.k_hi = k_hi
+        self.schedule_decay = schedule_decay
+        self.use_batch_objective = use_batch_objective
+        self.use_checkpoint_objective = use_checkpoint_objective
+        self.use_nesterov_momentum = use_nesterov_momentum
+
+    def create_state(self, index: int, weight: NDArray) -> EveState:
+        return EveState(weight)
+
+    def update(self, index: int, weight: NDArray, grad: NDArray, state: EveState):
+
+        assert isinstance(weight, NDArray)
+        assert isinstance(grad, NDArray)
+        lr = self._get_lr(index)
+        wd = self._get_wd(index)
+        self._update_count(index)
+
+        t = self._index_update_count[index]
+
+        # Preprocess grad
+        grad *= self.rescale_grad + wd * weight
```

Comment:
`wd` is the weight decay, no? If so it shouldn't be multiplied by the gradient, right?
But rather be something like:
```
grad = grad * self.rescale_grad + wd * weight
```
Or apply weight decay later after clipping.

Reply:
Good catch!  I took this nice compact line from MXNet's Nadam implementation which actually had that typo (`grad *= ...` vs `grad = grad * ...`).  For clipping, there's actually inconsistent ordering in MXNet.  For SGD, we have `rescaled_grad[row] = lr * rescale_grad * clip(grad[row], clip_gradient) + wd * weight[row]`.  For Adam, we have `rescaled_grad[row] = clip(grad[row] * rescale_grad + wd * weight[row], clip_gradient)`.  I'm using the order for Adam here.